TP11 Documentation and Release Notes Updates (#15065)
openshift/openshift-docs: servicemesh-install/topics/appendix-threescale-adapter.adoc

Prerequisites:

* {ProductName} 0.7.0+

> * A working 3scale account (https://www.3scale.net/signup/[SaaS] or https://access.redhat.com/documentation/en-us/red_hat_3scale_api_management/2.4/html/infrastructure/onpremises-installation[3scale 2.5 On-Premises])

Reviewer: The text for the link has been changed, however the URL is still going to 2.4. For 3scale 2.5, the structure has changed, so it's not as simple as just updating the number in the URL.

**Correct link:** https://access.redhat.com/documentation/en-us/red_hat_3scale_api_management/2.5/html/installing_3scale/onpremises-installation

Author: I've updated this link.

Pay particular attention to the `kind: handler` resource. You must update this with your 3scale credentials and the service ID of the API you want to manage.
====

. Modify the handler configuration with your 3scale configuration.
+
[source,yaml]
----
  apiVersion: "config.istio.io/v1alpha2"
  kind: handler
  metadata:
   name: threescale
   namespace:
  spec:
   adapter: threescale
   params:
     service_id: "<SERVICE_ID>"
     system_url: "https://<organization>-admin.3scale.net/"
     access_token: "<ACCESS_TOKEN>"
   connection:
     address: "threescale-istio-adapter:3333"
----

+
. Create the required resources:
+
----
oc create -f istio/ -n
----

== Generating custom resources

> The adapter includes a tool that allows you to generate the `handler`, `instance`, and `rule` custom resources.

Author: @dfennessy @PhilipGough @JStickler I changed the language here to clarify. Does this help?

Reviewer: LGTM

Reviewer 2: +1

Pay particular attention to the `kind: handler` resource. You must update this with your 3scale credentials and the service ID of the API you want to manage.
====

. Modify the handler configuration with your 3scale configuration.
+
[source,yaml]
----
  apiVersion: "config.istio.io/v1alpha2"
  kind: handler
  metadata:
   name: threescale
   namespace:
  spec:
   adapter: threescale
   params:
     service_id: "<SERVICE_ID>"
     system_url: "https://<organization>-admin.3scale.net/"
     access_token: "<ACCESS_TOKEN>"
   connection:
     address: "threescale-istio-adapter:3333"
----

+
. Create the required resources:
+
----
oc create -f istio/ -n 

> ----

Author: @dfennessy @PhilipGough I have removed the namespace label from the yaml example and the oc create command. Let me know if you're happy with this change.

Reviewer: Should it be `oc create -f istio/ -n {replace-me}`

Reviewer 2: Yes, needs a namespace provided as is or else to drop the `-n`. I prefer the former

Reviewer 2: @geekspertise `namespace: ` line should be removed entirely

Author: @PhilipGough and @dfennessy All namespace references are now removed.

> oc create -f istio/ -n istio-system

Reviewer 3: This will vary depending on where the controlplane is installed.

Reviewer 2: @brian-avery up until now the Maistra docs had been pointing towards `istio-system` namespace I believe?? If that has changed then good to update and add more details here also @dfennessy

Reviewer 3: With this release, we're adding support for multitenancy. In a single control plane mode everything will be in Istio-system as expected in previous releases. In multitenancy, the user can have as many control planes as they want and each one of the control planes has a different control plane namespace.

Author: @brian-avery @dfennessy I take it the 3scale team needs to factor multi-tenancy into their documentation.

Reviewer 2: Correct, we will need to update that upstream ASAP. I've created an issue

Reviewer 2: @dfennessy - IMO for now a note saying modify the namespace in the resources accordingly should be enough, wdyt?

Reviewer: Agreed @PhilipGough

Reviewer 4: I'd remove the namespace from the resource (`namespace: istio-system`) and just use the `-n` flag on the `oc` call.

Reviewer 5: +1 to Rob's suggestion, even in a single cluster wide installation it's not necessarily going to be in istio-system

Reviewer 2: +1 sounds good

Author: I've made these changes and updated the PR.

== Generating custom resources

> The adapter embeds a tool that allows generation of the `handler`, `instance`, and `rule` custom resources.

Reviewer 6: Where is the tool embedded?  In 3Scale?  Or in OpenShift? @PhilipGough @dfennessy

Reviewer 2: @JStickler its shipped with the adapter image so can be accessed locally if they pull the image or on OpenShift where the adapter is running

Author: I updated the language to clarify. See new content.

|File to save produced manifests to
|No
|Standard output

|`-v`
|Outputs the CLI version and exits
|No
|
|===

=== Generate templates from URL examples
This example generates the templates from the URL, service, and personal token:

----

> 3scale-gen-config --url="https://<organization>-admin.3scale.net:443" --service="123456789" --token="[redacted]"

Reviewer 7: Prefix commands that don't need root access with $ and ones that do need root access with #

Author: Fixed.

|File to save produced manifests to
|No
|Standard output

|`-v`
|Outputs the CLI version and exits
|No
|
|===

=== Generate templates from URL examples
This example generates the templates from the URL, service, and personal token:

----
3scale-gen-config --url="https://<organization>-admin.3scale.net:443" --service="123456789" --token="[redacted]"
----

This example generates the templates from the URL, service, and personal token but with a unique ID, that otherwise would have been derived from the URL:

----
3scale-gen-config --url="https://<organization>-admin.3scale.net" --uid="my-unique-id" --service="123456789" --token="[redacted]"
----

> The previous example might fail to pass validation if the given unique ID did contains special characters. You can use the fixup mode to correct that:

Reviewer 6: ```suggestion
The previous example might fail to pass validation if the given unique ID contains special characters. You can use the fixup option to attempt to correct validation errors:
```

Author: I had to add this change manually locally. I'll push it up here on the next PR update.

> The previous example might fail to pass validation if the given unique ID did contains  special characters. You can use the fixup mode to correct that:

Reviewer 3: double space between contains and special

Author: Fixed!

----

== Routing service traffic through the adapter

> To drive traffic for your service through the 3scale adapter, you need to match the rule `destination.labels["service-mesh.3scale.net"] == "true"` that you previously created in the configuration, in the `kind: rule resource`.

Reviewer 3: I think this is likely to be `kind: rule` as an example, see: https://istio.io/docs/concepts/policies-and-telemetry/#rules

Reviewer 2: Correct, that's the format upstream. Changing to `kind: rule` resource. should suffice.

Author: @PhilipGough I removed "resource" from the tickmarks.

= Authenticating requests
This Technology Preview release supports the following authentication methods:

* *Standard API Keys*: single randomized strings or hashes acting as an identifier and a secret token.

> * *Application identifier and key pairs*: immutable identifier and mutable secret key strings.

Reviewer 8: "OpenID authentication method" needs to be added here

Author: @vramosp and @dfennessy I've added text for "OpenID authentication method" per Vanessa's request. Please review.
